rfc/0006-kubewarden-versioning.md

```diff
@@ -0,0 +1,203 @@
+|:-------------|:--------------------------------|
+| Feature Name | Kubewarden versioning           |
+| Start Date   | Mar 3rd 2022                    |
+| Category     | Versioning                      |
+| RFC PR       | [fill this in after opening PR] |
+
+# Summary
+[summary]: #summary
+
+Kubewarden is composed of many core components that work together in
+order to build the final product. This is a proposal to harmonize
+Kubewarden versioning schema.
+
+# Motivation
+[motivation]: #motivation
+
+Having a way to specify what version of Kubewarden is being installed
+or executed would make it easier for the Kubewarden team to describe,
+market and support Kubewarden (as in user support).
+
+It would help users in understanding what version of Kubewarden they
+are about to install or execute.
+
+## Examples / User Stories
+[examples]: #examples
+
+> Users do not need to keep track of component versions and how they
+> work together. They just need to care about running a specific
+> Kubewarden version.
+
+> Users are able to easily understand that a new version of Kubewarden
+> is available, and that more than one component might be upgraded in
+> order to move to the new Kubewarden version.
+
+> A user has an issue with Kubewarden. Given its multiple component
+> nature, we would like to know what was the Kubewarden version that
+> got installed, and through what versions it was upgraded.
+
+# Detailed design
+[design]: #detailed-design
+
+## Many components, one experience
+
+Kubewarden core components, which are exposed to the user are:
+
+- Helm Charts
+- Kubewarden Controller
+- Policy Server
+- kwctl
+
+As a second dimension, there are also policies and their respective
+SDK's. This proposal is going to ignore this dimension in purpose,
+because Kubewarden runtime components (`policy-server` and `kwctl`)
+should be compatible with policies previously built, even with older
+language SDK's.
+
+### Public interfaces
+
+Our public interfaces with users are as follows:
+
+- Helm Charts
+  - `values.yaml`
+- Kubewarden Controller
+  - CLI args
+  - Custom Resource Definitions
+    - Versioned
+- Policy Server
+  - CLI args
+  - Configuration
+    - sigstore related files
+    - Usually handled by the Kubewarden Controller. Regardless we
+      should treat this as public interface, given some users might
+      run the Policy Server independent of the Kubewarden Controller,
+      and might configure it with the configuration file.
+- kwctl
+  - CLI args
+  - Configuration
+    - `sources.yaml`
+    - `docker.json`
+    - sigstore related files
+
+There is also the `waPC` contract between the `policy-evaluator`
+(consumed by the Policy Server and kwctl) and Policies and SDK's.
+
+## Kubewarden toplevel version
+
+This proposal envisions using [`CalVer`](https://calver.org/) to
+version the Kubewarden project.
+
+Had Kubewarden be released today, it would be `Kubewarden
+22.03`. This makes it trivial to understand how old your version of
```

Comment:
> For the first, I would prefer vulnerability scans and bills of materials.

I don't think they are mutually excluyent. I agree, but I think the information calver gives you is always valuable, besides security scans and BOM's.

>  For the later, I am undecided if I prefer semver. If I were a customer, I would prefer to know that I can update without non-backwards compat changes between releases, for example only updating minor versions. And schedule major versions for a later moment, as by the nature of Kubewarden, it may very well mean pausing the cluster workload.

Given there are a number of really big dependencies on our product, I don't think semver is really meaningful. I think WRT Kubewarden users what matters is "what is my upgrade path?". And to document that upgrade path (and their pre-requirements) I don't think semver offers any advantage over calver (I'm always referring to the umbrella version).

Reply:
My main concern here is users not knowing if the upgrade of the helm charts are painless or not.
I realize now that the helm charts have their own semver, and that carries that info.

I'm happy with calver in that aspect, resolving this conversation!